### PR TITLE
Run a tox thread per connection

### PR DESCRIPTION
**Pull request by suhr**

### CODE REVIEW

**Comment by suhr** [comment]
~~`SecretKey` needs an implementation of `FromStr`.~~ (done)

File: src/tox.rs [hunk 1]
```diff
@@ -515,15 +557,17 @@ fn tox_loop(request_rx: std::sync::mpsc::Receiver<Request>, mut answer_tx: Unbou
...
 }
 
-pub fn spawn_tox() -> ToxHandle {
+pub fn spawn_tox(secret_key: Option<rstox::core::SecretKey>) -> ToxHandle {
     use std::sync::mpsc;
 
     let (request_tx, request_rx) = mpsc::channel();
     let (answer_tx, answer_rx) = unbounded_channel();
+    let guard = ToxGuard::new();
+    let handle = guard.clone();
```
Comment (kpp) [inline src/tox.rs]: Nope. It won't work.

File: src/tox.rs [hunk 2]
```diff
@@ -12,9 +13,37 @@ const BOOTSTRAP_KEY: &'static str =
 
 const CLIENT_NAME: &'static str = "ws-client";
 
+#[derive(Clone)]
+pub struct ToxGuard {
+    is_dropped: Arc<atomic::AtomicBool>
+}
+
+impl ToxGuard {
```
Comment (kpp) [inline src/tox.rs:21]: Best practice: oneshot channel

Comment (kpp) [reply src/tox.rs:21]: https://subscription.packtpub.com/book/application_development/9781788623926/8/ch08lvl1sec75/using-the-oneshot-channel

Comment (suhr) [reply src/tox.rs:21]: It doesn't have that nice RAII semantics.

Comment (kpp) [reply src/tox.rs:21]: Still you call drop explicitly https://github.com/tox-rs/ws-tox/pull/22/files#diff-639fbc4ef05b315af92b4d836c31b023R98. You can call .send in that closure too.

Also I think https://docs.rs/futures/0.1.11/src/futures/sync/oneshot.rs.html#272 receiver is marked as finished if all of Senders are dropped.

Comment (suhr) [reply src/tox.rs:21]: I still don't like to use channel machinery for something as simple as a drop switch.

Comment (kpp) [reply src/tox.rs:21]: Fine.

File: src/main.rs [hunk 3]
```diff
@@ -24,72 +20,45 @@ where
...
 
-            // accept the request to be a ws connection if it does
             let f = upgrade
                 .accept()
                 .map_err(|e| IoError::new(IoErrorKind::Other,
                     format!("websocket accept err: {}", e)
                 ))
                 .and_then(move |(s, _h)| {
                     let (sink, stream) = s.split();
-                    let tox_tx = (*tox_tx.lock().unwrap()).clone();
```
Comment (kpp) [inline src/main.rs:88]: You have to send SK back to client.

Comment (suhr) [reply src/main.rs:88]: Shouldn't a client get everything with `Info` request?

Comment (suhr) [reply src/main.rs:88]: Oh wait, it doesn't make much sense to call `Info` on a blank profile. But still sending SK could be done in an another PR (it requires a new event in ws-tox-protocol).

**Review by kpp — APPROVED** [review]

**Pull request merged**